src/cobra/core/gene.py

```diff
@@ -299,3 +271,360 @@ def _repr_html_(self):
             n_reactions=len(self.reactions),
             reactions=format_long_string(", ".join(r.id for r in self.reactions), 200),
         )
+
+
+class GPR(Module):
+    """A Gene Reaction rule in a cobra model, using AST as base class.
+
+    Parameters
+    ----------
+    gpr_from : Expression or Module or AST
+        A GPR in AST format
+    """
+
+    def __init__(
+        self, gpr_from: Union[Expression, Module, AST] = None, *args, **kwargs
+    ):
+        super().__init__(*args, **kwargs)
+        self._genes = set()
+        if gpr_from:
+            if isinstance(gpr_from, str):
+                self.from_string(gpr_from)
+                raise TypeError(
+                    f"GPR accepts AST, not string. "
+                    f'Next time, use GPR().from_string("{gpr_from}")'
+                )
+            elif isinstance(gpr_from, (Expression, Module)):
+                cleaner = GPRCleaner()
+                cleaner.visit(gpr_from)
+                self._genes = cleaner.gene_set
```

Comment:
Possibly make a copy here?

```suggestion
                self._genes = cleaner.gene_set.copy()
```

Reply:
Doesn't make sense to me, since it is set. We need the cleaner to walk over the tree, but we don't need a copy.

Reply:
It's just an intuition right now, the way it currently stands, both the reaction and the cleaner hold a reference to the gene set. So I'm wondering if the cleaner can be garbage collected when it goes out of scope or if that is prevented due to there being a reference to one of its attributes. A copy would avoid that. I should test that, though.